src/SkeletalTrapezoidation.cpp

```diff
@@ -0,0 +1,3083 @@
+{
+    for (edge_t& edge : graph.edges)
+    {
+        if (!edge.prev)
+        {
+            edge.data.setMarked(false);
+            edge.twin->data.setMarked(false);
+        }
+    }
+}
+
+void SkeletalTrapezoidation::setBeadCount(const BeadingStrategy& beading_strategy)
+{
+    for (edge_t& edge : graph.edges)
+    {
+        if (edge.data.isMarked())
+        {
+            edge.to->data.bead_count = beading_strategy.optimal_bead_count(edge.to->data.distance_to_boundary * 2);
+        }
+    }
+
+    // fix bead count at locally maximal R
+    // also for marked regions!! See TODO s in generateTransitionEnd(.)
+    for (node_t& node : graph.nodes)
+    {
+        if (isLocalMaximum(node))
+        {
+            if (node.data.distance_to_boundary < 0)
+            {
+                RUN_ONCE(logWarning("Distance to boundary not yet computed for local maximum!\n"));
+                node.data.distance_to_boundary = std::numeric_limits<coord_t>::max();
+                bool first = true;
+                for (edge_t* edge = node.some_edge; first || edge != node.some_edge; edge = edge->twin->next)
+                {
+                    node.data.distance_to_boundary = std::min(node.data.distance_to_boundary, edge->to->data.distance_to_boundary + vSize(edge->from->p - edge->to->p));
+                }
+            }
+            coord_t bead_count = beading_strategy.optimal_bead_count(node.data.distance_to_boundary * 2);
+            node.data.bead_count = bead_count;
+        }
+    }
+}
+
+void SkeletalTrapezoidation:: filterUnmarkedRegions(const BeadingStrategy& beading_strategy)
+{
+    for (edge_t& edge : graph.edges)
+    {
+        if (!isEndOfMarking(edge))
+        {
+            continue;
+        }
+        assert(edge.to->data.bead_count >= 0 || edge.to->data.distance_to_boundary == 0);
+        coord_t max_dist = 400; // beading_strategy.getTransitioningLength(edge.to->data.bead_count)
+        filterUnmarkedRegions(&edge, edge.to->data.bead_count, 0, max_dist, beading_strategy);
+    }
+}
+
+bool SkeletalTrapezoidation::filterUnmarkedRegions(edge_t* to_edge, coord_t bead_count, coord_t traveled_dist, coord_t max_dist, const BeadingStrategy& beading_strategy)
+{
+    coord_t r = to_edge->to->data.distance_to_boundary;
+    bool dissolve = false;
+    for (edge_t* next_edge = to_edge->next; next_edge && next_edge != to_edge->twin; next_edge = next_edge->twin->next)
+    {
+        coord_t length = vSize(next_edge->to->p - next_edge->from->p);
+        if (next_edge->to->data.distance_to_boundary < r && !shorterThen(next_edge->to->p - next_edge->from->p, 10))
+        { // only walk upward
+            continue;
+        }
+        if (next_edge->to->data.bead_count == bead_count)
+        {
+            dissolve = true;
+        }
+        else if (next_edge->to->data.bead_count < 0)
+        {
+            dissolve = filterUnmarkedRegions(next_edge, bead_count, traveled_dist + length, max_dist, beading_strategy);
```

Comment:
Not once for every edge: Look at the return right before the scope ends on the for-loop. You get there every time you don't enter the continue on the top of the for loop.

This is in itself written a bit strangely...., it would be better if it just loops until it got the right iterator, then put the rest of that outside of the for loop (so you can even ditch one of the returns, since they're identical).

It's still unclear how much the recursion would actually hit even with the above taken into account though.